tests/integration/devfile/cmd_devfile_push_test.go

```diff
@@ -17,32 +17,46 @@ var _ = Describe("odo devfile push command tests", func() {
 	var namespace, context, cmpName, currentWorkingDirectory, projectDirPath string
 	var projectDir = "/projectDir"
 	var sourcePath = "/projects/nodejs-web-app"
+	var clusterType helper.CliRunner
 
-	// TODO: all oc commands in all devfile related test should get replaced by kubectl
-	// TODO: to goal is not to use "oc"
-	oc := helper.NewOcRunner("oc")
+	// Using program commmand according to cluter type in devfile
+	if os.Getenv("KUBERNETES") == "true" {
+		clusterType = helper.NewKubectlRunner("kubectl")
+	} else {
+		clusterType = helper.NewOcRunner("oc")
+	}
 
 	// This is run after every Spec (It)
 	var _ = BeforeEach(func() {
 		SetDefaultEventuallyTimeout(10 * time.Minute)
-		namespace = helper.CreateRandProject()
 		context = helper.CreateNewContext()
-		currentWorkingDirectory = helper.Getwd()
 		projectDirPath = context + projectDir
-		cmpName = helper.RandString(6)
-
-		helper.Chdir(context)
-
 		os.Setenv("GLOBALODOCONFIG", filepath.Join(context, "config.yaml"))
 
 		// Devfile push requires experimental mode to be set
 		helper.CmdShouldPass("odo", "preference", "set", "Experimental", "true")
+
+		if os.Getenv("KUBERNETES") == "true" {
+			homeDir := helper.GetUserHomeDir()
+			kubeConfigFile := helper.CopyKubeConfigFile(filepath.Join(homeDir, ".kube", "config"), filepath.Join(context, "config"))
```

Comment:
Why is kubeconfig copied only for Kubernetes?

Reply:
Comment https://github.com/openshift/odo/pull/2965#discussion_r413744746 elaborates the reason in a proper manner.

Reply:
Yes, I understand that. But the same problem will be with `helper.CreateRandProject` that calls `odo project create` that also modifies kubeconfig. Or am I missing something?

Reply:
Actually the issue was with running the test in parallel on kubernetes cluster. By default the project is not active when it is created through kubectl command on kubernetes cluster, a .lock file is created while set-context is used to make it active. In parallel run this .lock file causes the issue of ```[1] [kubectl] error: open (path)/k3s.yaml.lock: The file exists.``` as described in https://github.com/openshift/odo/pull/2965#pullrequestreview-399016104.

On the other hand in case of openshift cluster such thing does not happen when it creates the project and others using oc command. Yes ofcourse it modifies the kubeconfig but does not create a .lock file I guess.

Reply:
> On the other hand in case of openshift cluster such thing does not happen when it creates the project and others using oc command. Yes ofcourse it modifies the kubeconfig but does not create a .lock file I guess. Does it make sense?

`helper.CreateRandProject` is not using `oc`, but `odo` (don't know why I think that it should use `oc`).

But this doesn't mean that the problem with the parallel run is not there with OpenShift.
If  `helper.CreateRandProject` gets called multiple times at the same time, they will try to modify the same KUBECONFIG, and you will have the same problem as with `kubectl`. And in this situation, it is even more dangerous because odo doesn't use locking to prevent race condition when editing the KUBECONFIG

Reply:
> `helper.CreateRandProject` is not using `oc`, but `odo` (don't know why I think that it should use `oc`).

AFAIU `odo` is being used to create project because in all the test scenarios we are validating `odo project create` functionality while creating the project, so we should not care about how `oc new-project create` works in `odo` project. Ofcourse there are few functionality which we are borrowing from `oc` as there is no such options that `odo` command provides.

> But this doesn't mean that the problem with the parallel run is not there with OpenShift.
> If `helper.CreateRandProject` gets called multiple times at the same time, they will try to modify the same KUBECONFIG, and you will have the same problem as with `kubectl`. And in this situation, it is even more dangerous because odo doesn't use locking to prevent race condition when editing the KUBECONFIG

ok, however when I tested the same script on 4 test nodes I didn't see any failure from openshift side rather I could be able to see the ```.lock``` file error from kubernetes cluster side. 
I am not sure how these ```KUBECONFIG``` files are being handled during the creation of  random project against openshift and kubernetes cluster behind the scene respectively. Could you please add some more content like how it works, what call it makes etc... during project creation?
I will make the changes as per your suggestions.

Reply:
> ok, however when I tested the same script on 4 test nodes I didn't see any failure from openshift side rather I could be able to see the `.lock` file error from kubernetes cluster side.

Yes, because that problem is just hidden due to the fact that there is no locking mechanism to prevent two processes to modify the same file at the same time. But that doesn't mean that it is not happening. 
If you would have two test specs running at the same time and they would be executing a  command that uses active project/namespsce from KUBECONFIG one of them would be using the wrong one.

> Could you please add some more content like how it works, what call it makes etc... during project creation?

`odo project create`? It is not doing anything special. Just creates the project (the equivalent of `oc create project <name>`) and then sets it as active `odo project <name>`.
The setting is as active is where it can collide because it modifies KUBECONFIG.
So if the same KUBECONFIG is being used and the process tries to set active project at the same time. What is going to be written to KUBECONFIG?

Reply:
> If you would have two test specs running at the same time and they would be executing a command that uses active project/namespsce from KUBECONFIG one of them would be using the wrong one.
> odo project create? It is not doing anything special. Just creates the project (the equivalent of oc create project <name>) and then sets it as active odo project <name>. The setting is as active is where it can collide because it modifies KUBECONFIG.

@kadel hmm, but i wonder even on running test in parallel on 4 node against OpenShift cluster none of the time it throws any error related to KUBECONFIG or error like it throws against Kubernetes cluster. Is it a bug on OpenShift ? or OpenShift handles it in abetter way unlike Kubernetes ?

Reply:
> If you would have two test specs running at the same time and they would be executing a command that uses active project/namespsce from KUBECONFIG one of them would be using the wrong one.

@kadel I did a quick experiment to confirm the weird behavior but unfortunately none of the time I am able to hit the issue of KUBECONFIG collision against OpenShift cluster on 8 test node. 
Command executed :
```make test-cmd-project``` at least 5 times.
Experimented on:
```
func CreateRandProject() string {
  projectName := RandString(10)
  fmt.Fprintf(GinkgoWriter, "Creating a new project: %s\n", projectName)
  // session := CmdShouldPass("odo", "project", "create", projectName, "-w", "-v4")
  session := CmdShouldPass("oc", "new-project", projectName)
  // Expect(session).To(ContainSubstring("New project created"))
  Expect(session).To(ContainSubstring(projectName))
  return projectName
}
var _ = BeforeEach(func() {
    SetDefaultEventuallyTimeout(10 * time.Minute)
    SetDefaultConsistentlyDuration(30 * time.Second)
    context = helper.CreateNewContext()
    os.Setenv("GLOBALODOCONFIG", filepath.Join(context, "config.yaml"))
    project = helper.CreateRandProject()
  })
  // Clean up after the test
  // This is run after every Spec (It)
  var _ = AfterEach(func() {
    helper.DeleteProject(project)
    helper.DeleteDir(context)
    os.Unsetenv("GLOBALODOCONFIG")
  })
  Context("Machine readable output tests", func() {
    It("Help for odo project list should contain machine output", func() {
      output := helper.CmdShouldPass("odo", "project", "list", "--help")
      Expect(output).To(ContainSubstring("Specify output format, supported format: json"))
    })
    It("Help for odo project list should contain machine output", func() {
      output := helper.CmdShouldPass("odo", "project", "list", "--help")
      Expect(output).To(ContainSubstring("Specify output format, supported format: json"))
    })
    [similar spec 6 more times]
  })
```
Can we file an issue on OpenShift to confirm the behavior before going for a major changes in our repo ? cc_ @amitkrout

Reply:
Right, Lets check if there is any issue reported on ````OKD``` or ```oc```. Otherwise i would say lets drop a mail to ```aos-devel``` to get a clarity on this.

Reply:
> Can we file an issue on OpenShift to confirm the behavior before going for a major changes in our repo ? cc_ @amitkrout

It is not issued with OpenShift itself. I'm saying that the same problem as we have with namespace collisions in Kubernetes tests is also present in our OpenShift tests. The only reason why we don't see it is that in most of our tests we use `--project` flag.

We simply should treat Openshift and Kubernetes the same way. If we are doing something for Kubernetes it should be also done for OpenShift tests.

Reply:
> The only reason why we don't see it is that in most of our tests we use --project flag.

Ahh...Now i understood
It's the ```--project``` flag which is saving us from the failure. using oc command it was passing because I was using freshly created project for each spec in script i pasted above to test it with oc which @amitkrout somehow missed it.
Actually i am in the favor of test script in sync for both the cluster but i was little confused why it was even not failing with oc.
Thanks for the info. Will update it ASAP.

Reply:
> It's the `--project` flag which is saving us from the failure. using oc command it was passing because I was using freshly created project for each spec in script i pasted above to test it with oc which @amitkrout somehow missed it.

yes, exactly.  And you never know when someone adds test that doesn't use `--project` flag. When that happens we would start getting flakes due to wrong project being used.